OSGI Missing Provide-Capability entry for org.apache.shiro.crypto.hash.HashSpi

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/shiro/issues?q=is%3Aissue) and found no similar issues.


### Environment

Apache Karaf 4.4.4 & JDK 17 & Fedora 38

### Shiro version

2.0.0 alpha 3

### What was the actual outcome?

During password encryption I can see that shiro isn't able to see argon hash provider.

### What was the expected outcome?

Shiro is able to use bcrypt & argon hash providers

### How to reproduce

Import shiro bundles:
    <feature version="${org.apache.shiro.version}">shiro-core</feature>
    <bundle>mvn:org.apache.shiro.crypto/shiro-hashes-bcrypt/${org.apache.shiro.version}</bundle>
    <bundle>mvn:org.apache.shiro.crypto/shiro-hashes-argon2/${org.apache.shiro.version}</bundle>

Deploy any application that uses shiro and is able to run under Karaf & OSGI

### Debug logs

Caused by: java.lang.UnsupportedOperationException: Cannot create a hash with the given algorithm: argon2id
	at org.apache.shiro.crypto.hash.DefaultHashService.computeHash(DefaultHashService.java:98) ~[!/:2.0.0-alpha-3]
	at org.apache.shiro.authc.credential.DefaultPasswordService.passwordsMatch(DefaultPasswordService.java:181) ~[!/:2.0.0-alpha-3]
	at tech.pantheon.chronos.aaa.impl.security.realm.impl.TokenCredentialsMatcher.doCredentialsMatch(TokenCredentialsMatcher.java:36) ~[?:?]
	at org.apache.shiro.realm.AuthenticatingRealm.assertCredentialsMatch(AuthenticatingRealm.java:600) ~[!/:2.0.0-alpha-3]
	at org.apache.shiro.realm.AuthenticatingRealm.getAuthenticationInfo(AuthenticatingRealm.java:581) ~[!/:2.0.0-alpha-3]
	at Proxy7d5c2798_ad1d_48c5_aa34_93f64a110877.getAuthenticationInfo(Unknown Source) ~[?:?]
	at org.apache.shiro.authc.pam.ModularRealmAuthenticator.doSingleRealmAuthentication(ModularRealmAuthenticator.java:180) ~[!/:2.0.0-alpha-3]
	at org.apache.shiro.authc.pam.ModularRealmAuthenticator.doAuthenticate(ModularRealmAuthenticator.java:273) ~[!/:2.0.0-alpha-3]
	at org.apache.shiro.authc.AbstractAuthenticator.authenticate(AbstractAuthenticator.java:198) ~[!/:2.0.0-alpha-3]

--- a/crypto/hash/pom.xml
+++ b/crypto/hash/pom.xml
@@ -49,6 +49,14 @@
                             org.apache.shiro.lang*;version="${shiro.osgi.importRange}",
                             *
                         </Import-Package>
+                        <Require-Capability>
+                          osgi.extender; filter:="(osgi.extender=osgi.serviceloader.registrar)",                         
+                          osgi.extender; filter:="(osgi.extender=osgi.serviceloader.processor)",
+                          osgi.serviceloader; filter:="(osgi.serviceloader=org.apache.shiro.crypto.hash.HashSpi)"; cardinality:=multiple
+                        </Require-Capability>
+                        <Provide-Capability>
+                          osgi.serviceloader; osgi.serviceloader=org.apache.shiro.crypto.hash.HashSpi
+                        </Provide-Capability>
                         <_removeheaders>Bnd-LastModified</_removeheaders>
                         <_reproducible>true</_reproducible>
                     </instructions>

--- a/crypto/support/hashes/argon2/pom.xml
+++ b/crypto/support/hashes/argon2/pom.xml
@@ -54,13 +54,19 @@
                 <configuration>
                     <instructions>
                         <Bundle-SymbolicName>org.apache.shiro.hashes.argon2</Bundle-SymbolicName>
-                        <Export-Package>org.apache.hashes.argon2*;version=${project.version}</Export-Package>
+                        <Export-Package>org.apache.shiro.crypto.support.hashes.argon2*;version=${project.version}</Export-Package>
                         <Import-Package>
                             org.apache.shiro*;version="${shiro.osgi.importRange}",
                             org.aopalliance*;version="[1.0.0, 2.0.0)",
                             com.google.inject*;version="1.3",
                             *
                         </Import-Package>
+                        <Require-Capability>
+                          osgi.extender; filter:="(osgi.extender=osgi.serviceloader.registrar)"                         
+                        </Require-Capability>                        
+                        <Provide-Capability>
+                          osgi.serviceloader; osgi.serviceloader=org.apache.shiro.crypto.hash.HashSpi
+                        </Provide-Capability>                        
                         <_removeheaders>Bnd-LastModified</_removeheaders>
                         <_reproducible>true</_reproducible>
                     </instructions>

--- a/crypto/support/hashes/bcrypt/pom.xml
+++ b/crypto/support/hashes/bcrypt/pom.xml
@@ -54,13 +54,19 @@
                 <configuration>
                     <instructions>
                         <Bundle-SymbolicName>org.apache.shiro.hashes.bcrypt</Bundle-SymbolicName>
-                        <Export-Package>org.apache.hashes.bcrypt*;version=${project.version}</Export-Package>
+                        <Export-Package>org.apache.shiro.crypto.support.hashes.bcrypt*;version=${project.version}</Export-Package>
                         <Import-Package>
                             org.apache.shiro*;version="${shiro.osgi.importRange}",
                             org.aopalliance*;version="[1.0.0, 2.0.0)",
                             com.google.inject*;version="1.3",
                             *
                         </Import-Package>
+                        <Require-Capability>
+                          osgi.extender; filter:="(osgi.extender=osgi.serviceloader.registrar)"                         
+                        </Require-Capability>                        
+                        <Provide-Capability>
+                          osgi.serviceloader; osgi.serviceloader=org.apache.shiro.crypto.hash.HashSpi
+                        </Provide-Capability>                                  
                         <_removeheaders>Bnd-LastModified</_removeheaders>
                         <_reproducible>true</_reproducible>
                     </instructions>

--- a/support/features/src/main/resources/features.xml
+++ b/support/features/src/main/resources/features.xml
@@ -22,10 +22,12 @@
     <!-- Shiro core features: -->
 
     <feature name="shiro-core" version="${project.version}">
+        <feature version="[1,2)">spifly</feature>
         <bundle dependency="true">mvn:commons-beanutils/commons-beanutils/${commons.beanutils.version}</bundle>
         <bundle dependency="true">mvn:org.apache.commons/commons-configuration2/${commons.configuration2.version}</bundle>
         <bundle dependency="true">mvn:commons-collections/commons-collections/${commons.collection.version}</bundle>
         <bundle dependency="true">mvn:org.apache.commons/commons-text/${commons.text.version}</bundle>
+        <bundle>mvn:org.apache.geronimo.specs/geronimo-annotation_1.3_spec/1.1</bundle>
         <bundle>mvn:org.apache.shiro/shiro-lang/${project.version}</bundle>
         <bundle>mvn:org.apache.shiro/shiro-cache/${project.version}</bundle>
         <bundle>mvn:org.apache.shiro/shiro-config-ogdl/${project.version}</bundle>
@@ -67,7 +69,6 @@
     <feature name="shiro-hazelcast" version="${project.version}">
         <feature version="${project.version}">shiro-core</feature>
         <bundle>mvn:org.apache.geronimo.specs/geronimo-jta_1.1_spec/1.1.1</bundle>
-        <bundle>mvn:org.apache.geronimo.specs/geronimo-annotation_1.3_spec/1.1</bundle>
         <bundle>mvn:com.hazelcast/hazelcast-all/${hazelcast.version}</bundle>
         <bundle>mvn:org.apache.shiro/shiro-hazelcast/${project.version}</bundle>
     </feature>

